File: chp01_Singleton/Swift/Testing_DatabaseConnection_thread_safety.swift
import Foundation
import XCTest

class DatabaseConnection {
    private static var _sharedInstance: DatabaseConnection?
    
    internal static func _initializeInstance() {
        _sharedInstance = DatabaseConnection()
    }

    static var sharedInstance: DatabaseConnection {
        guard let instance = _sharedInstance else {
            _initializeInstance(); return _sharedInstance!
        }
        return instance
    }
    
    private init() {}
    
    func connectToDatabase() {
        // Connect to the database here
        print("Connected to database")
    }
    
    func executeQuery(query: String) {
        // Execute the query on the database
        print("Executing query: \(query)")
    }
    
    static func resetInstance() {
        _sharedInstance = nil
        print("DatabaseConnection instance reset for testing purposes")
    }
}

// Test class using Dependency Injection
class DatabaseConnectionTest {
    private let databaseConnection: DatabaseConnection
    
    init(databaseConnection: DatabaseConnection) {
        self.databaseConnection = databaseConnection
    }
    
    func runTest() {
        // Example test case: connect to database and execute a query
        databaseConnection.connectToDatabase()
        databaseConnection.executeQuery(query: "SELECT * FROM test_table")
    }
}

// Thread Safety Test using XCTest
class DatabaseConnectionThreadSafetyTest: XCTestCase {
    func testSingletonThreadSafety() {
        let numThreads = 10
        let expectation = XCTestExpectation(description: "All threads should complete")
        expectation.expectedFulfillmentCount = numThreads
        var instances: [DatabaseConnection] = []
        let instanceLock = NSLock()
        
        DispatchQueue.concurrentPerform(iterations: numThreads) { i in
            let instance = DatabaseConnection.sharedInstance
            instanceLock.lock()
            instances.append(instance)
            instanceLock.unlock()
            expectation.fulfill()
        }
        
        wait(for: [expectation], timeout: 10.0)
        
        // Verify that all instances are the same
        let firstInstance = instances.first
        for instance in instances {
            XCTAssert(firstInstance === instance, "All instances should be the same")
        }
    }
}

// Usage
let dbConnection = DatabaseConnection.sharedInstance
dbConnection.connectToDatabase()
dbConnection.executeQuery(query: "SELECT * FROM users")

// Dependency Injection Example
let dbTest = DatabaseConnectionTest(databaseConnection: dbConnection)
dbTest.runTest()

// Reset Singleton Instance for Testing Purposes
DatabaseConnection.resetInstance()
// Reinitialize to prevent crash
DatabaseConnection._initializeInstance()
let dbConnection2 = DatabaseConnection.sharedInstance
dbConnection2.connectToDatabase()
dbConnection2.executeQuery(query: "SELECT * FROM new_users")

File: chp01_Singleton/Kotlin/Testing_DatabaseConnection_thread_safety.kt
import org.testng.annotations.Test
import org.testng.Assert.assertSame
import org.testng.Assert.assertNotNull
import kotlin.concurrent.thread
import java.util.concurrent.CountDownLatch

object DatabaseConnection {
    @Volatile
    private var instance: DatabaseConnection? = null

    init {
        // Initialize the database connection here
        println("Database connection initialized")
    }

    fun executeQuery(query: String): String {
        // Execute the query and return the result
        println("Executing query: $query")
        return "Query executed successfully"
    }

    @Synchronized
    fun resetInstance() {
        instance = null
        println("Database connection instance reset for testing purposes")
    }
}

// Test class using Dependency Injection
class DatabaseConnectionTest(private val databaseConnection: DatabaseConnection) {
    fun runTest() {
        // Example test case: execute a query
        val result = databaseConnection.executeQuery("SELECT * FROM test_table")
        println("Test Query Result: $result")
    }
}

import org.testng.annotations.BeforeMethod

class DatabaseConnectionThreadSafetyTest {
    @BeforeMethod
    fun setup() {
        DatabaseConnection.resetInstance()
    }
    @Test
    fun testSingletonInstanceWithMultipleThreads() {
        val latch = CountDownLatch(10)
        val instances = mutableListOf<DatabaseConnection>()

        for (i in 1..10) {
            thread {
                val instance = DatabaseConnection
                synchronized(instances) {
                    instances.add(instance)
                }
                latch.countDown()
            }
        }

        latch.await()

        val firstInstance = instances.first()
        instances.forEach { instance ->
            assertSame("All instances should be the same", firstInstance, instance)
        }
        assertNotNull("Singleton instance should not be null", firstInstance)
    }
}

fun main() {
    val dbInstance = DatabaseConnection
    val result = dbInstance.executeQuery("SELECT * FROM users")
    println(result)

    // Dependency Injection Example
    val dbTest = DatabaseConnectionTest(dbInstance)
    dbTest.runTest()

    // Reset the Singleton Instance for Testing Purposes
    DatabaseConnection.resetInstance()
    val dbInstance2 = DatabaseConnection
    val result2 = dbInstance2.executeQuery("SELECT * FROM new_users")
    println(result2)
}

// Gradle commands for executing the application and test:
// To run the main application, make sure you have Gradle installed and use the following command:
// gradle run
// To run the tests, make sure you have TestNG configured and use the following command:
// gradle test --info
// If Gradle wrapper is available, you can use ./gradlew run or ./gradlew test --info
